packages/cubejs-backend-shared/src/time.ts

```diff
@@ -26,10 +31,113 @@ export const TIME_SERIES: Record<string, (range: DateRange, timestampPrecision:
     .map(d => [d.format(`YYYY-MM-DDT00:00:00.${'0'.repeat(digits)}`), d.endOf('quarter').format(`YYYY-MM-DDT23:59:59.${'9'.repeat(digits)}`)]),
 };
 
-type TimeSeriesOptions = {
-  timestampPrecision: number
+/**
+ * Parse PostgreSQL-like interval string into object
+ * E.g. '2 years 15 months 100 weeks 99 hours 15 seconds'
+ * Negative units are also supported
+ * E.g. '-2 months 5 days -10 hours'
+ */
+export function parseSqlInterval(intervalStr: SqlInterval): ParsedInterval {
+  const interval: ParsedInterval = {};
+  const parts = intervalStr.split(/\s+/);
+
+  for (let i = 0; i < parts.length; i += 2) {
+    const value = parseInt(parts[i], 10);
+    const unit = parts[i + 1];
+
+    // Remove ending 's' (e.g., 'days' -> 'day')
+    const singularUnit = (unit.endsWith('s') ? unit.slice(0, -1) : unit) as unitOfTime.DurationConstructor;
```

Comment:
But it doesn't validate any errors with granularities. We can make regexp more strict. For example 

`/^(-?\d+) (second|minute|hour|day|week|month|quarter|year)s?$/`

Or validate it here with string, because you are casting `string` to `unitOfTime.DurationConstructor`.

Additional: Could you add a test with incorrect input for granularity unit?

Reply:
There is no need to validate anything here as the interval comes from the cube configuration definition that was validated during the parsing/init phase.